File: Src/java/elm/src/main/java/org/cqframework/cql/elm/visiting/FunctionalElmVisitor.java
package org.cqframework.cql.elm.visiting;

import java.util.Objects;
import java.util.function.BiFunction;
import org.cqframework.cql.elm.tracking.Trackable;

/**
 * The is a base class for visitors that apply functions to all the visited ELM elements.
 * Useful for quick visitor implementations, such as counting all nodes, or finding a specific element
 * type.
 */
public class FunctionalElmVisitor<T, C> extends BaseElmLibraryVisitor<T, C> {

    private final BiFunction<Trackable, C, T> defaultResult;
    private final BiFunction<T, T, T> aggregateResult;

    /**
     * Constructor that takes a default visit function and an aggregate result function.
     * @param defaultResult the function for processing a visited element
     * @param aggregateResult the function for aggregating results
     */
    public FunctionalElmVisitor(BiFunction<Trackable, C, T> defaultResult, BiFunction<T, T, T> aggregateResult) {
        this.defaultResult = Objects.requireNonNull(defaultResult);
        this.aggregateResult = Objects.requireNonNull(aggregateResult);
    }

    @Override
    public T defaultResult(Trackable elm, C context) {
        return this.defaultResult.apply(elm, context);
    }

    @Override
    public T aggregateResult(T aggregate, T nextResult) {
        return this.aggregateResult.apply(aggregate, nextResult);
    }

    public static <C, T> FunctionalElmVisitor<T, C> from(
            BiFunction<Trackable, C, T> defaultResult, BiFunction<T, T, T> aggregateResult) {
        return new FunctionalElmVisitor<>(defaultResult, aggregateResult);
    }
}

File: Src/java/elm/src/main/kotlin/org/cqframework/cql/elm/visiting/FunctionalElmVisitor.kt
package org.cqframework.cql.elm.visiting

import java.util.function.BiFunction
import org.hl7.elm.r1.Element

/**
 * This is a base class for visitors that apply functions to all the visited ELM elements. Useful
 * for quick visitor implementations, such as counting all nodes, or finding a specific element
 * type.
 */
class FunctionalElmVisitor<T, C>(
    private val defaultResult: BiFunction<Element, C, T>,
    private val aggregateResult: BiFunction<T, T, T>
) : BaseElmLibraryVisitor<T, C>() {

    public override fun defaultResult(elm: Element, context: C): T {
        return defaultResult.apply(elm, context)
    }

    public override fun aggregateResult(aggregate: T, nextResult: T): T {
        return aggregateResult.apply(aggregate, nextResult)
    }

    companion object {
        fun <C, T> from(
            defaultResult: BiFunction<Element, C, T>,
            aggregateResult: BiFunction<T, T, T>
        ): FunctionalElmVisitor<T, C> {
            return FunctionalElmVisitor(defaultResult, aggregateResult)
        }

        fun <C, T> from(defaultResult: BiFunction<Element, C, T>): FunctionalElmVisitor<T, C> {
            return from(defaultResult) { _: T, b: T -> b }
        }
    }
}
